### PR TITLE
Fix #212, simplify build loop

### LINKED ISSUE #212: `lorri ping_` doesn’t trigger reevaluation if the daemon watches the nix file already
I noticed that if I run `lorri direnv` (via the direnv integration), the daemon sometimes does not start a build.

More specifically, if I remove the `shell_gc_root` (which the daemon prints in its `Started` messages), the environment won’t be rebuild if I run `lorri ping_ /path/to/shell.nix`, as long as the daemon is still running. If I restart the daemon, it is rebuilt after `ping_`.

The ping is received by the daemon here:

https://github.com/target/lorri/blob/a0ff0a51f742f1e73f4adefd0d46c67c6032016c/src/ops/daemon.rs#L42-L44
it is written to `accept_messages_tx` and read from the reading end further down:

https://github.com/target/lorri/blob/a0ff0a51f742f1e73f4adefd0d46c67c6032016c/src/ops/daemon.rs#L60-L74

The problem lies in `daemon.add()`, which has a different semantics than what I thought when writing this code:

https://github.com/target/lorri/blob/a0ff0a51f742f1e73f4adefd0d46c67c6032016c/src/daemon.rs#L64-L80

It adds projects to the daemon, but only if the project was not added yet (`or_insert_with`).

Instead, we need to be able to influence the `build_loop`s somehow, which are running inside threads. I propose to add a channel going from the daemon to each `build_loop`s and extending the hash map by the `tx` going to each builder:

https://github.com/target/lorri/blob/a0ff0a51f742f1e73f4adefd0d46c67c6032016c/src/daemon.rs#L29-L31

Then we can forward a message to the build_loop that a build should be queried, even if no file has changed in the meantime.

### PR DESCRIPTION
**Pull request by curiousleo**
This is a better fix for #212 than my previous "minimal change" fix in https://github.com/target/lorri/pull/214. It builds on top of https://github.com/target/lorri/pull/217.

**What's in this PR?**
- fixes #212 based on this idea: https://github.com/target/lorri/pull/214#issuecomment-553963285
- simplifies the build loop in `BuildLoop::forever`
- replaces our own debouncing logic by `notify`'s (but keeping all the tests to ensure we get the same behaviour)

**Changes to `BuildLoop::forever`**
This PR changes both what the build loop does and how it does it.

The behavioural change is from
```
on notify {
    if notification_relevant() {
        rebuild()
    }
}
```
to
```
on notify {
    if notification_relevant() {
        rebuild()
    }
}
on ping {
    if !gc_root.exists() {
        rebuild()
    }
}
```
That is, a rebuild is now _also_ triggered if no GC root exists when a ping is received. This fixes #212.

The implementation also changes: Previously the main loop blocked on `self.watch.wait_for_change()`. Now it uses `chan::select!` to wait on a message, which comes either from a ping or from a notification.

**Using `notify`'s built-in debouncing logic**
Switching from [`Watcher::new_immediate`](https://docs.rs/notify/5.0.0-pre.1/notify/trait.Watcher.html#tymethod.new_immediate) to [`Watcher::new`](https://docs.rs/notify/5.0.0-pre.1/notify/trait.Watcher.html#tymethod.new) lets us use `notify`'s own debouncing logic. In combination with the switch from timout-based, blocking calls inside the build loop to a message-based approach, this allows for a substantial simplification of `Watch` as well as the removal of `FilterTimeoutIterator`.

### CODE REVIEW

**Comment by curiousleo** [comment]
PTAL

File: src/build_loop.rs [hunk 1]
```diff
@@ -63,43 +63,65 @@ impl<'a> BuildLoop<'a> {
...
 
-        send(Event::Started(Reason::ProjectAdded(
+        // The project has just been added, so run the builder in the first iteration
+        let mut reason = Some(Event::Started(Reason::ProjectAdded(
             self.project.nix_file.clone(),
         )));
+        let mut output_paths = None;
+
+        // Drain pings initially: we're going to trigger a first build anyway
+        rx_ping.try_iter().for_each(drop);
```
Comment (grahamc) [inline src/build_loop.rs]: Is it possible this will drain pings for other projects, too?

Comment (curiousleo) [reply src/build_loop.rs]: No: `rx_ping` is created by `Daemon::add`, so there is one channel per project. For each project, there is a `Handler` in `Daemon::handler_threads` which contains the sender `tx` corresponding to the `rx_ping` receiver. It's somewhat implicit, but I hope that this structure makes it clear that it is _intended_ to have one of those channels per project.

**Comment by curiousleo** [comment]
PTAL

**Pull request merged**